carbonserver/render.go

```diff
@@ -744,33 +737,56 @@ func (listener *CarbonserverListener) Render(req *protov2.MultiFetchRequest, str
 		key, size := listener.getRenderCacheKeyAndSize(targets, format.String()+"grpc")
 		var res interface{}
 		cacheT0 := time.Now()
-		res, fromCache, err = getWithCache(logger, listener.queryCache, key, size, 60,
-			func() (interface{}, error) {
-				return fetchAndStreamMetricsFunc(true)
-			})
-		tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
-		if err == nil {
-			listener.prometheus.cacheRequest("query", fromCache)
-			if fromCache {
-				atomic.AddUint64(&listener.metrics.QueryCacheHit, 1)
-				cacheResponse := res.([]response)
+
+		item := listener.queryCache.getQueryItem(key, size, 60)
+		res, cacheOk := item.FetchOrLock()
+		switch {
+		case !cacheOk:
+			atomic.AddUint64(&listener.metrics.QueryCacheMiss, 1)
+			// TODO: should chan buffer size be configurable?
+			responseChan := make(chan response, 1000)
+			err = fetchMetricsFunc(responseChan)
+			if err != nil {
+				item.StoreAbort()
+				tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
+			} else {
 				go func() {
-					for _, r := range cacheResponse {
-						responseChan <- r
+					var fullResponse []response
```

Comment:
nit: maybe `responses` would be simpler, `full` phrasing could be a bit confusing, like if there is full then there must be partial which makes you wonder and check further, while `responses` is clear than it's just array of `response`s

Reply:
Changed to `responses`